source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -235,13 +235,15 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
     decoding_state_.setCompleteBodyAvailable(true);
   }
 
-  if (!decoding_state_.sendHeaders()) {
-    ENVOY_LOG(trace, "decodeHeaders: Skipped");
-    return FilterHeadersStatus::Continue;
+  FilterHeadersStatus status = FilterHeadersStatus::Continue;
+  if (decoding_state_.sendHeaders()) {
+    status = onHeaders(decoding_state_, headers, end_stream);
+    ENVOY_LOG(trace, "onHeaders returning {}", static_cast<int>(status));
   }
 
-  const auto status = onHeaders(decoding_state_, headers, end_stream);
-  ENVOY_LOG(trace, "decodeHeaders returning {}", static_cast<int>(status));
+  if (decoding_state_.shouldRemoveContentLength()) {
```

Comment:
In the case of the content-length needs to be removed, maybe call removeContentlength() before calling onHeaders() to avoid send the content-length header to the external processing server?

Reply:
I think 
1) Whether external_processor calculate the content-length header or not is out of ext_proc control .They won't base on this to decide their logic. Their behavior is also outside of our concern.
Also, In buffer mode + header processing, they will re-calculate body length anyway.

2) I think it doesn't hurt to convey this information to external_processor.

Reply:
Yeah, functionality wise, it does not hurt. 

Performance wise, it cause some unnecessary processing in both ext_proc server and Envoy.

For example, if body_mode_ is STREAMED or BUFFERED_PARTIAL, and sending content-length to the ext_proc server, they might mutate body, and recalculate the content-length, and sending back to Envoy in the header mutation. However, Envoy simply ignored it.   So if Envoy does not have content-length sent to the ext_proc server, ext_proc server won't do any of this in the first place.

Reply:
I think external processor can still send back the content length header no matter if it is presented or not.

This is out of ext_proc control and should not be controlled by ext_proc

Reply:
In the case of shouldRemoveContentLength() is true, envoy is removing content-length from Envoy side, and ignoring content-length mutation event if it is added by the ext_proc server.  It's the circumstances that envoy will force to do chunk-encoding. Thus if we define the API docs clearly says ext_proc server should only update content-length, not adding. And Envoy do not send content-length to ext_proc server in those circumstances, it can avoid ext_proc do some useless work.
Any way, my 2 cents.

Reply:
It's probably a good idea to forward as much original context (including headers) as possible. This is useful for example in security audit or if you're trying to fingerprint a client.

Reply:
Ok, originally I am thinking ext_proc server doesn't have the processing mode configuration as Envoy ext_proc filter to decide whether to send back a modified content-length.  However, they might be able to derive that based on the header and body they received to make that decision.

Reply:
BTW, onHeaders() call openStream() which might return error. Do we need to consider that before call removeContentLength()?

Reply:
I think it should not matter because:
1. onHeaders() is for header processing. It is independent of body processing which is what removal content_length  is really for 
2. here is removing the content length in original request, it is independent of onHeaders that is for sidestream processing.
3. I think `StopIteration` is returned when onHeaders returns error OR external processor has closed the stream

Reply:
SGTM